core/primitives-core/src/types.rs

```diff
@@ -23,7 +29,77 @@ pub type ShardId = u64;
 /// Balance is type for storing amounts of tokens.
 pub type Balance = u128;
 /// Gas is a type for storing amount of gas.
-pub type Gas = u64;
+#[derive(
+    Debug,
+    Default,
+    BorshSerialize,
+    BorshDeserialize,
+    Serialize,
+    Deserialize,
+    Clone,
+    Copy,
+    Hash,
+    PartialEq,
+    Eq,
+    PartialOrd,
+    Ord,
+)]
+pub struct Gas(u64);
+
+impl Gas {
+    pub fn new(val: u64) -> Self {
+        Self(val)
+    }
+
+    pub fn saturating_add(self, rhs: Self) -> Self {
+        Self(self.0.saturating_add(rhs.0))
+    }
+
+    pub fn checked_add(self, rhs: Self) -> Option<Self> {
+        self.0.checked_add(rhs.0).map(Self::new)
+    }
+
+    pub fn wrapping_add(self, rhs: Self) -> Self {
+        Self(self.0.wrapping_add(rhs.0))
+    }
+
+    pub fn saturating_sub(self, rhs: Self) -> Self {
+        Self(self.0.saturating_sub(rhs.0))
+    }
+
+    pub fn checked_sub(self, rhs: Self) -> Option<Self> {
+        self.0.checked_sub(rhs.0).map(Self::new)
+    }
+
+    pub fn wrapping_sub(self, rhs: Self) -> Self {
+        Self(self.0.wrapping_sub(rhs.0))
+    }
+
+    pub fn get(&self) -> u64 {
+        self.0
+    }
```

Comment:
This should probably not exist. Looking at the uses, there are couple common situations:

* Divide gas into portions (`impl ops::Div<u64> for Gas { type Result = Gas; }` and same for `Rem`);
  * or add an inherent method...
* Divide gas by gas (`impl ops::Div<Gas> for Gas { type Result = u64; }`);
  * or add an inherent method, note though that the result is unitless.
* Subtraction/addition/etc – use one of the methods defined above. Same for the multiplication and such;
* (Parts of) complex computations like in `compute_new_gas_price` seems like could be inherent methods on the type.

Reply:
Good point.  Agreed.  My current plan is to get the code base compiling using this method for now and then see what the common patterns are that can be factored out.